### PR TITLE
2 -  In memory stores for party record and user management to support update operations [DPP-1198]

### PR DESCRIPTION
**Pull request by pbatko-da**
changelog_begin
changelog_end

Depends on https://github.com/digital-asset/daml/pull/14793

(Extracted from https://github.com/digital-asset/daml/pull/14623/)

### CODE REVIEW

File: ledger/participant-integration-api/src/main/scala/platform/usermanagement/PersistentUserManagementStore.scala [hunk 1]
```diff
@@ -99,30 +100,36 @@ class PersistentUserManagementStore(
...
   override def createUser(
       user: domain.User,
       rights: Set[domain.UserRight],
-  )(implicit loggingContext: LoggingContext): Future[Result[Unit]] = {
+  )(implicit loggingContext: LoggingContext): Future[Result[User]] = {
     inTransaction(_.createUser) { implicit connection: Connection =>
       withoutUser(user.id) {
         val now = epochMicroseconds()
-        val internalId = backend.createUser(user, createdAt = now)(connection)
+        val dbUser = UserManagementStorageBackend.DbUserPayload(
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/main/scala/platform/usermanagement/PersistentUserManagementStore.scala:116]: Don't we need some TODOs around here. This implementation is not finished

- no resourceVersion check/enforcement
- no storage of annotations

File: ledger/participant-integration-api/src/test/lib/scala/platform/store/interning/MockStringInterning.scala [hunk 2]
```diff
@@ -65,7 +65,7 @@ class MockStringInterning extends StringInterning {
         rawStringInterning.tryExternalize(id).map(Party.assertFromString)
     }
 
-  private[store] def reset(): Unit = synchronized {
+  def reset(): Unit = synchronized {
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/test/lib/scala/platform/store/interning/MockStringInterning.scala:68]: Oh that one rears its ugly head again

File: ledger/participant-integration-api/src/test/lib/scala/platform/store/platform/partymanagement/ParticipantPartyStoreTests.scala [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+              ledgerPartyExists = _ => Future.successful(false),
+            )
+            _ = res1.left.value shouldBe ParticipantPartyRecordStore.PartyNotFound(party)
+          } yield succeed
+        }
+      }
+    }
+  }
+
+}
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/test/lib/scala/platform/store/platform/partymanagement/ParticipantPartyStoreTests.scala:281]: Missing test for incorrect resourceVersion in an update

File: ledger/participant-integration-api/src/test/lib/scala/platform/store/platform/usermanagement/UserManagementStoreTests.scala [hunk 4]
```diff
@@ -267,4 +317,125 @@ trait UserManagementStoreTests extends TestResourceContext with Matchers with Ei
...
+              ),
+            )
+          )
+          _ = res1.left.value shouldBe UserManagementStore.UserNotFound(userId)
+        } yield succeed
+      }
+    }
+
+  }
+
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/test/lib/scala/platform/store/platform/usermanagement/UserManagementStoreTests.scala:440]: Missing test for incorrect resourceVersion in an update

File: ledger/participant-state-index/src/main/scala/com/daml/ledger/participant/state/index/impl/inmemory/InMemoryParticipantPartyRecordStore.scala [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+    state.put(party, updatedInfo)
+    updatedInfo
+  }
+
+  private def doCreatePartyRecord(
+      partyRecord: ParticipantParty.PartyRecord
+  ): PartyRecordInfo = {
+    val info = PartyRecordInfo(
+      party = partyRecord.party,
+      resourceVersion = 0,
```
Comment (mziolekda) [inline ledger/participant-state-index/src/main/scala/com/daml/ledger/participant/state/index/impl/inmemory/InMemoryParticipantPartyRecordStore.scala:157]: There is no guarantee that the client will set it correctly to 0 in partyRecord. Shouldn't we verify it?

File: ledger/participant-state-index/src/main/scala/com/daml/ledger/participant/state/index/impl/inmemory/InMemoryParticipantPartyRecordStore.scala [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+  ): PartyRecordInfo = {
+    val existingAnnotations = info.annotations
+    val updatedAnnotations =
+      partyRecordUpdate.metadataUpdate.annotationsUpdateO.fold(existingAnnotations) {
+        case AnnotationsUpdate.Merge(newAnnotations) => existingAnnotations.concat(newAnnotations)
+        case AnnotationsUpdate.Replace(newAnnotations) => newAnnotations
+      }
+    val updatedInfo = PartyRecordInfo(
+      party = party,
+      resourceVersion = info.resourceVersion + 1,
```
Comment (mziolekda) [inline ledger/participant-state-index/src/main/scala/com/daml/ledger/participant/state/index/impl/inmemory/InMemoryParticipantPartyRecordStore.scala:145]: There is no gurarantee partyRecordUpdate has the correct resourceVersion

**Review by mziolekda — COMMENTED** [review]
Left a couple of comments

File: ledger/participant-integration-api/src/main/scala/platform/store/backend/common/UserManagementStorageBackendImpl.scala [hunk 7]
```diff
@@ -39,54 +36,64 @@ object UserManagementStorageBackendImpl extends UserManagementStorageBackend {
...
     internalId.get
   }
 
   override def getUser(
       id: UserId
-  )(connection: Connection): Option[UserManagementStorageBackend.DbUser] = {
+  )(connection: Connection): Option[UserManagementStorageBackend.DbUserWithId] = {
     SQL"""
-       SELECT internal_id, user_id, primary_party, created_at
+       SELECT internal_id, user_id, primary_party,  created_at
```
[the remark this answers is not in the document]
Comment (pbatko-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/backend/common/UserManagementStorageBackendImpl.scala:57]: Reverting

File: ledger/participant-state-index/src/main/scala/com/daml/ledger/participant/state/index/v2/UserManagementStore.scala [hunk 8]
```diff
@@ -9,6 +9,58 @@ import com.daml.logging.LoggingContext
...
 import scala.concurrent.{ExecutionContext, Future}
 
+case class UserUpdate(
+    id: Ref.UserId,
+    primaryPartyUpdateO: Option[Option[Ref.Party]] = None,
+    isDeactivatedUpdateO: Option[Boolean] = None,
+    metadataUpdate: ObjectMetaUpdate,
+)
+
+case class PartyRecordUpdate(
```
[the remark this answers is not in the document]
Comment (pbatko-da) [reply ledger/participant-state-index/src/main/scala/com/daml/ledger/participant/state/index/v2/UserManagementStore.scala:19]: Moving

File: ledger/participant-state-index/src/main/scala/com/daml/ledger/participant/state/index/v2/UserManagementStore.scala [hunk 9]
```diff
@@ -9,6 +9,58 @@ import com.daml.logging.LoggingContext
...
+    def annotations: Map[String, String] = nonEmptyAnnotations
+  }
+
+  object Merge {
+    def apply(annotations: Map[String, String]): Option[Merge] = {
+      if (annotations.isEmpty) None
+      else Some(new Merge(annotations))
+    }
+
+    def fromNonEmpty(annotations: Map[String, String]): Merge = {
```
[the remark this answers is not in the document]
Comment (pbatko-da) [reply ledger/participant-state-index/src/main/scala/com/daml/ledger/participant/state/index/v2/UserManagementStore.scala:40]: Annotations will be validated and only then processed to produce e.g. 'Merge' instances. I want it to be clear that valid 'Merge' instances must have non-empty annotations and I'm relying on a suggestive method name 'fromNonEmpty' coding errors.

File: ledger/participant-state-index/src/main/scala/com/daml/ledger/participant/state/index/v2/ParticipantPartyRecordStore.scala [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+  type Result[T] = Either[Error, T]
+
+  final case object PartyRecordNotFoundOnUpdateException extends RuntimeException
+
+  sealed trait Error extends RuntimeException
+
+  final case class PartyNotFound(party: Ref.Party) extends Error
+  final case class PartyRecordNotFound(party: Ref.Party) extends Error
+  final case class PartyRecordExists(party: Ref.Party) extends Error
+  final case class ConcurrentPartyUpdate(party: Ref.Party) extends Error
```
[the remark this answers is not in the document]
Comment (pbatko-da) [reply ledger/participant-state-index/src/main/scala/com/daml/ledger/participant/state/index/v2/ParticipantPartyRecordStore.scala:46]: Ok

File: ledger/participant-state-index/src/main/scala/com/daml/ledger/participant/state/index/impl/inmemory/InMemoryUserManagementStore.scala [hunk 11]
```diff
@@ -32,11 +36,47 @@ class InMemoryUserManagementStore(createAdmin: Boolean = true) extends UserManag
 
   override def createUser(user: User, rights: Set[UserRight])(implicit
       loggingContext: LoggingContext
-  ): Future[Result[Unit]] =
+  ): Future[Result[User]] =
     withoutUser(user.id) {
-      state.update(user.id, UserInfo(user, rights))
+      val userWithResourceVersion =
+        user.copy(metadata = user.metadata.copy(resourceVersionO = Some("0")))
```
[the remark this answers is not in the document]
Comment (pbatko-da) [reply ledger/participant-state-index/src/main/scala/com/daml/ledger/participant/state/index/impl/inmemory/InMemoryUserManagementStore.scala:42]: I'm planning to validate at the `ApiUserManagamentService` layer. Do you find it useful to add it here as well?

[the remark this answers is not in the document]
Comment (mziolekda) [reply ledger/participant-state-index/src/main/scala/com/daml/ledger/participant/state/index/impl/inmemory/InMemoryUserManagementStore.scala:42]: in that case, I rest my case. We have decided to put all the internal message consistency tests in the ledger api layer, so we don't need to repeat ourselves here

File: ledger/participant-integration-api/src/main/scala/platform/usermanagement/PersistentUserManagementStore.scala (hunk 1 above)
Comment (pbatko-da) [reply ledger/participant-integration-api/src/main/scala/platform/usermanagement/PersistentUserManagementStore.scala:116]: That's already included in the next PR: https://github.com/digital-asset/daml/pull/14834/files

File: ledger/participant-state-index/src/main/scala/com/daml/ledger/participant/state/index/impl/inmemory/InMemoryUserManagementStore.scala [hunk 12]
```diff
@@ -32,11 +36,47 @@ class InMemoryUserManagementStore(createAdmin: Boolean = true) extends UserManag
...
+          case AnnotationsUpdate.Replace(newAnnotations) => newAnnotations
+        }
+      val newResourceVersion = (userInfo.user.metadata.resourceVersionO
+        // TODO um-for-hub: Use error codes
+        .getOrElse(
+          sys.error(
+            s"Could not find resource version on user: ${userInfo.user}. All created users must have a resource version"
+          )
+        )
+        .toLong + 1).toString
```
[the remark this answers is not in the document]
Comment (pbatko-da) [reply ledger/participant-state-index/src/main/scala/com/daml/ledger/participant/state/index/impl/inmemory/InMemoryUserManagementStore.scala:65]: Good catch, fixing

File: ledger/participant-state-index/src/main/scala/com/daml/ledger/participant/state/index/impl/inmemory/InMemoryParticipantPartyRecordStore.scala (hunk 5 above)
Comment (pbatko-da) [reply ledger/participant-state-index/src/main/scala/com/daml/ledger/participant/state/index/impl/inmemory/InMemoryParticipantPartyRecordStore.scala:157]: That will be checked at the API service layer as part of the request validation

File: ledger/participant-state-index/src/main/scala/com/daml/ledger/participant/state/index/impl/inmemory/InMemoryParticipantPartyRecordStore.scala (hunk 6 above)
Comment (pbatko-da) [reply ledger/participant-state-index/src/main/scala/com/daml/ledger/participant/state/index/impl/inmemory/InMemoryParticipantPartyRecordStore.scala:145]: Fixing

File: ledger/participant-integration-api/src/test/lib/scala/platform/store/interning/MockStringInterning.scala (hunk 2 above)
Comment (pbatko-da) [reply ledger/participant-integration-api/src/test/lib/scala/platform/store/interning/MockStringInterning.scala:68]: Ah, it must have reappeared during rebasing/merging and conflict resolution that I was doing. Reverting

File: ledger/participant-integration-api/src/test/lib/scala/platform/store/platform/partymanagement/ParticipantPartyStoreTests.scala (hunk 3 above)
Comment (pbatko-da) [reply ledger/participant-integration-api/src/test/lib/scala/platform/store/platform/partymanagement/ParticipantPartyStoreTests.scala:306]: Added

File: ledger/participant-integration-api/src/main/scala/platform/usermanagement/PersistentUserManagementStore.scala [hunk 13]
```diff
@@ -221,9 +236,29 @@ class PersistentUserManagementStore(
...
+    )
+  }
+
+  private def toDomainUser(
+      dbUser: UserManagementStorageBackend.DbUserPayload
+  ): domain.User = {
+    val payload = dbUser
+    domain.User(
+      id = payload.id,
+      primaryParty = payload.primaryPartyO,
```
[the remark this answers is not in the document]
Comment (pbatko-da) [reply ledger/participant-integration-api/src/main/scala/platform/usermanagement/PersistentUserManagementStore.scala:253]: It's handled in the next PR: https://github.com/digital-asset/daml/pull/14834/files#r957519274

File: ledger/participant-state-index/src/main/scala/com/daml/ledger/participant/state/index/impl/inmemory/InMemoryParticipantPartyRecordStore.scala [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+  ): Result[PartyRecordInfo] = {
+    val existingAnnotations = info.annotations
+    val updatedAnnotations =
+      partyRecordUpdate.metadataUpdate.annotationsUpdateO.fold(existingAnnotations) {
+        case AnnotationsUpdate.Merge(newAnnotations) => existingAnnotations.concat(newAnnotations)
+        case AnnotationsUpdate.Replace(newAnnotations) => newAnnotations
+      }
+    val currentResourceVersion = info.resourceVersion
+    val newResourceVersionEither = partyRecordUpdate.metadataUpdate.resourceVersionO match {
+      case None => Right(currentResourceVersion + 1)
```
Comment (mziolekda) [inline ledger/participant-state-index/src/main/scala/com/daml/ledger/participant/state/index/impl/inmemory/InMemoryParticipantPartyRecordStore.scala:147]: Right, I have re-read the API proposal that indeed has provisions for this field having been omitted.

**Review by mziolekda — APPROVED** [review]

File: ledger/participant-state-index/src/main/scala/com/daml/ledger/participant/state/index/impl/inmemory/InMemoryParticipantPartyRecordStore.scala [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+  override def createPartyRecord(
+      partyRecord: ParticipantParty.PartyRecord
+  )(implicit loggingContext: LoggingContext): Future[Result[ParticipantParty.PartyRecord]] = {
+    withState(withoutPartyRecord(partyRecord.party) {
+      val info = doCreatePartyRecord(partyRecord)
+      toPartyRecord(info)
+    })
+  }
+
+  override def updatePartyRecord(
```
[the remark this answers is not in the document]
Comment (pbatko-da) [reply ledger/participant-state-index/src/main/scala/com/daml/ledger/participant/state/index/impl/inmemory/InMemoryParticipantPartyRecordStore.scala:71]: Ok, noted

**Pull request merged**